Streaming query documentation (#231)
AxonIQ/reference-guide: axon-framework/queries/query-dispatchers.md

> Regardless whether you choose to use the `QueryBus` or the `QueryGateway`, both provide several types of queries. Axon Framework makes a distinction between three types, being:

Reviewer: ```suggestion
Regardless whether you choose to use the `QueryBus` or the `QueryGateway`, both provide several types of queries. Axon Framework makes a distinction between four types, being:
```

   Since we subscribed to updates with the `println()` method, the update will be printed out once it is received.

> ### Streaming queries

Reviewer: I'd expect the streaming query to be added to the top of this file in the "The Query Bus and Query Gateway".
We summarize the other types, so why not add the streaming query there too?

> The Streaming query allows a client to stream large database result sets. The Streaming query relies on reactive stream model, specifically the `Flux` type. Using streaming query, does not imply that reactive db drivers need to be used, users can hook to the stream and control the flow of data.

Reviewer: I am not overly certain about the last sentence in this paragraph.
Other changes I made here, are consistently using "streaming" instead of "Streaming".

```suggestion
The streaming query allows a client to, for example, stream large database result sets. The streaming query relies on the reactive stream model, specifically the `Flux` type. Using the streaming query does not imply that reactive database drivers should be used. Users can hook to the stream and control the flow of data.
```

> Streaming query is flexible enough to be used with **any** query return type. That means that any return type that is not Flux will automatically be converted to Flux, based on that Flux will return single or multiple items.

Reviewer: ```suggestion
The streaming query is flexible enough to use **any** query return type. That means that any return type that is not a `Flux` will automatically be converted to `Flux`. The `Flux` will return single or multiple items based on that.
```

> Query gateway provides nifty `streamingQuery` method to utilize streaming query easily. 
> It's simple to use, and requires just two parameters: query and expected response type class.
> `streamingQuery` method **is lazy**, meaning query is sent once Flux is subscribed to.

Reviewer: ```suggestion
The `QueryGateway` provides the `streamingQuery` method to utilize the streaming query. 
It's simple to use and requires just two parameters: the query payload and the expected response type class.
Note that the `streamingQuery` method **is lazy**, meaning the query is sent once the `Flux` is subscribed to.
```

> It is also possible to utilize streaming query with `query` method: `queryGateway.query(query,  ResponseTypes.fluxOf(CardSummary.class))`.
> However, this approach is **eager** and less intuitive as it requires more boilerplate code.
> In this case `query` method will send query immediately and return CompletableFuture which returns a payload of type Flux, that needs to be subscribed to.

Reviewer: To further clarify what you mean with "the query method," linking to that section in the documentation is helpful:

```suggestion
It is also possible to utilize the streaming query with the  [Point-to-Point query](query-dispatchers.md#point-to-point-queries) method: `queryGateway.query(query,  ResponseTypes.fluxOf(CardSummary.class))`.
However, this approach is **eager** and less intuitive as it requires more boilerplate code.
In this case, the `query` method will send the query immediately and return a `CompletableFuture`.
This `CompletableFuture` returns a payload of type `Flux` that you subscribe to.
```

Author: this is outdated, and will be removed, but will commit suggestion for now

> ```java

Reviewer: This sample comes completely out of the dark. 
A short (e.g. one or two sentences) introduction describing what's to come would be helpful to the readers.

@QueryHandler
public List<CardSummary> handle(FetchCardSummariesQuery query) {
        ...
    return cardRepository.findAll(); //1
}
        ...

public Flux<CardSummary> consumer() {
        return queryGateway.streamingQuery(query, CardSummary.class); //2
}
```

> 1. We are querying the `cardRepository` for all the cards, that can potently return a large result set containing thousands of items.
> 2. We are using the `queryGateway` to issue the query. If we used `multipleInstanceOf(CardSummary.class)` we would get large list transferred as single result message over the network, potentially causing a buffer overflow and max message size violation.
> Instead, we are using `streamingQuery(query, CardSummary.class)`, that will convert our response to Flux stream and chunk the result set into smaller messages of CardSummary.

Reviewer: ```suggestion
1. We are querying the `cardRepository` for all the cards. The repository can potentially return a result set containing thousands of items.
2. We are using the `queryGateway` to issue the query. If we used `multipleInstanceOf(CardSummary.class)`, we would get an extensive list transferred as a single result message over the network. This result can potentially cause a buffer overflow or maximum message size violation.
Instead of the multiple-instance-of approach, we use the `streamingQuery(query, CardSummary.class)`. This method will convert our response to a stream and chunk the result into smaller messages containing the `CardSummary` instances.
```

> Natively, if we want fine-grained control of the producing stream we can use Flux as return type:

Reviewer: ```suggestion
Natively, if we want fine-grained control of the producing stream, we can use `Flux` as the return type:
```

```java
@QueryHandler
public Flux<CardSummary> handle(FetchCardSummariesQuery query) {
        ...
    return reactiveCardRepository.findAll(); 
}
```

> When using Flux as return type, we can control backpressure, stream cancellation and implement more complex features like pagination.

Reviewer: ```suggestion
When using a `Flux` as the return type, we can control backpressure, stream cancellation and implement more complex features like pagination.
```

> > **Note**

Reviewer: It is better to have "named notes," like with the one directly coming after this one. So, something like "Axon Server Version considerations".
Furthermore, it looks a bit messy to have two notes directly after one another. 
Hence I'd recommend to move this note to a different portion of this subsection.

>
> Backpressure and stream cancellation features are only available with Axon Server 4.6 + version. Streaming queries used with pre Axon Server 4.6 version will work, but without these important features.
>

> > **Transaction leaking**
> >
> > Once consumer of streaming query receives Flux to subscribe to, transaction will be considered completed successfully. That means, that any subsequent messages on the stream will not be part of transaction. 
> > This includes errors, and as transaction is already over, error will not be propagated to parent transaction to invoke any rollback method.
> > This has implication that streaming query should not be used within Unit Of Work (within message handlers or any other transactional methods) to chaining other transactional actions (like sending a command or query).
> >

Reviewer: Next to the suggestion, I'd also link to the Unit of Work page in the reference guide for added clarity.

```suggestion
> **Transaction Leaking Concerns**
>
> Once a consumer of the streaming query receives the `Flux` to subscribe to, the transaction will be considered completed successfully. 
> That means that any subsequent messages on the stream will not be part of the transaction, including errors.
> As the transaction is already over an error will not be propagated to the parent transaction to invoke any rollback method.
> This has the implication that the streaming query should not be used within a Unit Of Work (within message handlers or any other transactional methods) to chain other transactional actions (like sending a command or query).
```

> #### Backpressure
> Backpressure is important feature in reactive systems that allows consumer to control the flow of data, and not to be overwhelmed by the producer.
> The streaming query implements pull-based back-pressure strategy, which means that the producer will emit data when consumer is ready to receive it.

Reviewer: I'd adjust the title slightly, mainly to further strengthen it's not just any back-pressure, but back-pressure for the streaming query solution:

```suggestion
#### Streaming Back-pressure

Back-pressure is an essential feature in reactive systems that allows consumers to control the data flow, ensuring they are not overwhelmed by the producer.
The streaming query implements a pull-based back-pressure strategy, which means that the producer will emit data when the consumer is ready to receive it.
```

> Under the hood, backpressure does `Hop to Hop` signal propagation (see below) and inherits gRPC's [HTTP2-based backpressure model](https://developers.google.com/web/fundamentals/performance/http2/#flow_control).

Reviewer: Although I get the focus on Axon Server here, the Reference Guide should not assume Axon Server is used whenever we're on the Axon Framework section.
I would thus adjust the third sentence somewhat to make it clear this is an afterthought.
Or, we could move Axon Server specifics into a separate block describing the approach when in a distributed environment.

At any rate, you would need to introduce the assumption that you're talking about Axon Server, instead of dropping it on the reader as a given.

As a result, backpressure will not behave intuitively and will not propagate exact request signal from consumer to producer. 
HTTP/2 and Netty flow control has internal buffers based on message size, Axon Framework and Axon Server prefetch messages into internal buffers based on message count. 
Result is that the producer will send number of messages until it fills all the buffers, then backpressure will kick in.

> **Hop to hop**
>
>Back-pressure signal is propagated per hop which is not an end to end connection,
> which allows intermediate Axon Server to handle backpressure between two connections
> and pre-fetch additional messages to increase overall performance.
>

> #### Cancellation
> The Streaming query can be implemented as infinitive stream, so it's important to cancel it when the client is not interested in receiving any more data.
> Use operators that can cancel stream to signal producer that you are not interested in receiving more data.

Reviewer: I don't fully comprehend what you intend with the second sentence starting with "use operators."
Hence, I'll refrain from suggesting a change there until you've provided some guidance on what the intent is.

```suggestion
#### Cancellation

The streaming query can be implemented as an infinitive stream.
Hence it's important to cancel it when the client is not interested in receiving any more data.
Use operators that can cancel stream to signal producer that you are not interested in receiving more data.
```

> It's important to note that similar to backpressure, cancellation signal is per hop, meaning it's propagated over the network to Axon Server and then to producer, which can produce some latency in stream cancellation.

Reviewer: Alright, I spot that we're again focused on Axon Server, as does the following `Error Handling` section.

A similar point should be made here: we can't assume users use Axon Server in the Axon Framework section of the Reference Guide.
Hence we need to introduce that fact.
Furthermore, some guidance on these three topics **without** Axon Server should be present too, if applicable.

Instead, we are using `streamingQuery(query, CardSummary.class)`, that will convert our response to Flux stream and chunk the result set into smaller messages of CardSummary.

Natively, if we want fine-grained control of the producing stream we can use Flux as return type:

```java
@QueryHandler
public Flux<CardSummary> handle(FetchCardSummariesQuery query) {
        ...
    return reactiveCardRepository.findAll(); 
}
```

When using Flux as return type, we can control backpressure, stream cancellation and implement more complex features like pagination.

> **Transaction leaking**
>
> Once consumer of streaming query receives Flux to subscribe to, transaction will be considered completed successfully. That means, that any subsequent messages on the stream will not be part of transaction. 
> This includes errors, and as transaction is already over, error will not be propagated to parent transaction to invoke any rollback method.
> This has implication that streaming query should not be used within Unit Of Work (within message handlers or any other transactional methods) to chaining other transactional actions (like sending a command or query).
>

#### Streaming back-pressure

> Back-pressure (flow control) is an essential feature in reactive systems that allows consumers to control the data flow, ensuring they are not overwhelmed by the producer.
> The streaming query implements a pull-based back-pressure strategy, which means that the producer will emit data when the consumer is ready to receive it.
> 
> If you are using Axon Server, for more information see flow control [documentation](../../axon-server/performance/flow-control.md).

Reviewer: ```suggestion
Back-pressure (flow control) is an essential feature in reactive systems that allows consumers to control the data flow, ensuring they are not overwhelmed by the producer.
The streaming query implements a pull-based back-pressure strategy, which means that the producer will emit data when the consumer is ready to receive it.

If you are using Axon Server, for more information see the [flow control documentation](../../axon-server/performance/flow-control.md).
```

#### Cancellation

> The streaming query can be implemented as an infinitive stream.
> Hence, it's important to cancel it once the client is not interested in receiving any more data.
> 
> 
> ```java
> public Flux<CardSummary> consumer() {
>         return queryGateway.streamingQuery(query, CardSummary.class)
>             .take(100)
>             .takeUntil(message -> somePredicate.test(message));
> }
> ```
> 
> Example above shows usage of `take` operators to limit the number of items to be emitted.

Reviewer: ```suggestion
The streaming query can be implemented as an infinitive stream.
Hence, it's important to cancel it once the client is not interested in receiving any more data.

The following sample shows how this could be achieved:

```java
public Flux<CardSummary> consumer() {
        return queryGateway.streamingQuery(query, CardSummary.class)
                           .take(100)
                           .takeUntil(message -> somePredicate.test(message));
}
\```

The example above shows how the `take` operator limits the number of items to be emitted.
```

#### Error handling

> Producer that produces an error by calling `onError(Throwable)` will terminate the handler execution and the consumer will have its `onError(Throwable)` subscription handler called as expected.
> 
> Exceptions do not flow upstream (from consumer to producer). 
> If an error happens on consumer side, consumer error will trigger cancel signal that will be propagated to producer, and effectively cancel the stream, without producer knowing the reason.
> 
> It's recommended to set a timeout on a query handler side in case of finite stream to protect against malfunctioning consumer or producer.

Reviewer: ```suggestion
A producer that produces an error by calling `onError(Throwable)` will terminate the handler execution.
The consumer will, in turn, have its `onError(Throwable)` subscription handler called.

Note that exceptions do not flow upstream (from consumer to producer). 
If an error happens on the consumer side, the consumer error will trigger a cancel signal propagated to the producer.
This signal will effectively cancel the stream without the producer knowing the reason.

Hence it's recommended to set a timeout on the query handler's side in case of a finite stream.
Essentially to protect against malfunctioning consumers or producers.
```

```java
@QueryHandler
public Flux<CardSummary> handle(FetchCardSummariesQuery query) {
...

> return reactiveCardRepository.findAll().timeout(Duration.ofSeconds(5));

Reviewer: ```suggestion
    return reactiveCardRepository.findAll().timeout(Duration.ofSeconds(5));
```

}
```

> Example above shows usage of `timeout` operator to cancel a request of response has not observed during 5 second span.

Reviewer: ```suggestion
The example above shows how the `timeout` operator is used to cancel a request if no responses have been observed during a five-second timespan.
```

> **Can streaming query replace subscription query?**

Reviewer: Is this intended as another subsection? Or is this intended as a notice (like the "reactor dependency" notice below)?

Author: I will remove it, its not that useful
